src/frontend/src/app/flow/canvas/canvas.component.ts

```diff
@@ -232,50 +214,67 @@ export class CanvasComponent implements AfterViewInit, OnDestroy {
     nodeMap: Map<string, Node>
   ): void {
     for (const key of Object.keys(pipeline)) {
-      if (key !== '$') {
-        pipeline[key].forEach((element: any) => {
-          const nodeInfo = this.getNodeInfo(element.$);
-          let node;
-
-          if (key === 'Exit') {
-            node = new Exit(
-              nodeInfo.id,
-              nodeInfo.name,
-              key,
-              nodeInfo.top,
-              nodeInfo.left
-            );
-          } else {
-            node = new Pipe(
-              nodeInfo.id,
-              nodeInfo.name,
-              key,
-              nodeInfo.top,
-              nodeInfo.left
-            );
-
-            graph.setNode(nodeInfo.name, {
-              label: nodeInfo.name,
-              shape: 'rect',
-              width: 200,
-              height: 100,
-            });
-
-            if (element.Forward) {
-              element.Forward.forEach((pipeNode: { $: { path: string } }) => {
-                forwards.push(new Forward(nodeInfo.name, pipeNode.$.path));
-              });
-            }
-          }
+      const nodeInfo = pipeline[key];
+      let node;
+
+      let x = 0;
+      let y = 0;
 
-          nodeMap.set(nodeInfo.name, node);
+      nodeInfo.attributes.forEach((attr: any) => {
+        if (attr.x) {
+          x = attr.x;
+        } else if (attr.y) {
+          y = attr.y;
+        }
+      });
+
+      node = new Pipe(nodeInfo.name, nodeInfo.name, nodeInfo.type, y, x);
+
+      graph.setNode(nodeInfo.name, {
+        label: nodeInfo.name,
+        shape: 'rect',
+        width: 200,
+        height: 100,
+      });
 
-          this.nodeService.addDynamicNode(node);
+      if (nodeInfo.forwards) {
+        nodeInfo.forwards.forEach((forward: any) => {
+          forward.attributes.forEach((attr: any) => {
+            if (attr.path) {
+              forwards.push(new Forward(nodeInfo.name, attr.path));
+            }
+          });
         });
       }
+
+      nodeMap.set(nodeInfo.name, node);
+
+      this.nodeService.addDynamicNode(node);
     }
   }
 
+  generateExits(exits: any[]): void {
+    exits.forEach((nodeInfo) => {
+      let x = 0;
+      let y = 0;
+      let path = '';
+
+      nodeInfo.attributes.forEach((attr: any) => {
+        if (attr.x) {
+          x = attr.x;
+        } else if (attr.y) {
+          y = attr.y;
+        } else if (attr.path) {
+          path = attr.path;
+        }
+      });
```

Comment:
You could apply the `test = undefined ?? 0` (=0) tactic to reduce this down to 3 lines.

Reply:
This was in some other places as well.

Reply:
Excuse me its an if else.